asab/storage/elasticsearch.py

```diff
@@ -0,0 +1,177 @@
+	}
+)
+
+
+class StorageService(StorageServiceABC):
+	"""
+	Depends on `aiohttp`.
+	"""
+
+	def __init__(self, app, service_name, config_section_name='asab:storage'):
+		super().__init__(app, service_name)
+		self.Loop = app.Loop
+
+		self.ESURL = asab.Config.get(config_section_name, 'elasticsearch_url')
+		# self._timeout = asab.Config.get(config_section_name, 'elasticsearch_timeout')
+
+		username = asab.Config.get(config_section_name, 'elasticsearch_username')
+		password = asab.Config.get(config_section_name, 'elasticsearch_password')
+
+		if username == '':
+			self._auth = None
+		else:
+			self._auth = aiohttp.BasicAuth(login=username, password=password)
+
+		self._ClientSession = None
+
+	async def finalize(self, app):
+		if self._ClientSession is not None and not self._ClientSession.closed:
+			await self._ClientSession.close()
+			self._ClientSession = None
+
+	def session(self):
+		if self._ClientSession is None:
+			self._ClientSession = aiohttp.ClientSession(auth=self._auth, loop=self.Loop)
+		elif self._ClientSession.closed:
+			self._ClientSession = aiohttp.ClientSession(auth=self._auth, loop=self.Loop)
+		return self._ClientSession
+
+	async def delete(self, index, _id=None):
+		if _id:
+			url = "{}{}/_doc/{}".format(self.ESURL, index, _id)
+		else:
+			url = "{}{}".format(self.ESURL, index)
+		async with self.session().request(method="DELETE", url=url) as resp:
+			assert resp.status == 200, "Unexpected response code: {}".format(resp.status)
+			resp = await resp.json()
+			if resp["acknowledged"]:
+				return resp
+			assert resp["result"] == "deleted", "Document was not deleted"
+			return resp
+
+	async def get_by(self, collection: str, key: str, value):
+		raise NotImplementedError("get_by")
+
+	async def get(self, index: str, obj_id) -> dict:
+		url = "{}{}/_doc/{}".format(self.ESURL, index, obj_id)
+		async with self.session().request(method="GET", url=url) as resp:
+			obj = await resp.json()
+
+		# Manipulate the output so that it is a requested object
+		ret = obj['_source']
+		ret['_v'] = obj['_version']
+		ret['_id'] = obj['_id']
+
+		return ret
+
+	def upsertor(self, index: str, obj_id=None, version: int = 0):
+		return ElasicSearchUpsertor(self, index, obj_id, version)
+
+	async def list(self, index, size=10000):
+		'''
+		Custom ElasticSearch method
+		'''
+		url = "{}{}/_search?size={}".format(self.ESURL, index, size)
+		async with self.session().request(method="GET", url=url) as resp:
+			assert resp.status == 200, "Unexpected response code: {}".format(resp.status)
+			return await resp.json()
+
+	async def indices(self, search_string=None):
+		'''
+		Custom ElasticSearch method
+		'''
+		url = "{}_cat/indices/{}?format=json".format(self.ESURL, search_string)
+		async with self.session().request(method="GET", url=url) as resp:
+			assert resp.status == 200, "Unexpected response code: {}".format(resp.status)
+			return await resp.json()
+
+	async def empty_index(self, index):
+		'''
+		Custom ElasticSearch method
```

Comment:
What is "empty_index" ... name of the method should contain **clear** noun -> is is "clear_index" or "create_index" ?

Reply:
Right, it is create index, will rename